Sources/SystemInternals/Exports.swift

```diff
@@ -36,47 +50,129 @@ public var system_errno: CInt {
 #endif
 
 public func system_open(_ path: UnsafePointer<CChar>, _ oflag: Int32) -> CInt {
+#if os(Windows)
+  var fh: CInt = -1
+  _ = _sopen_s(&fh, path, oflag, _SH_DENYNO, _S_IREAD | _S_IWRITE)
+  return fh
+#else
   open(path, oflag)
+#endif
 }
 
+#if os(Windows)
+public func system_open(
+  _ path: UnsafePointer<CChar>, _ oflag: Int32, _ mode: CInt
+) -> CInt {
+  var fh: CInt = -1
+  _ = _sopen_s(&fh, path, oflag, _SH_DENYNO, mode)
+  return fh
+}
+#else
 public func system_open(
   _ path: UnsafePointer<CChar>, _ oflag: Int32, _ mode: mode_t
 ) -> CInt {
   open(path, oflag, mode)
 }
+#endif
 
 public func system_close(_ fd: Int32) -> Int32 {
-  close(fd)
+#if os(Windows)
+  return _close(fd)
+#else
+  return close(fd)
+#endif
 }
 
+#if os(Windows)
+public func system_read(
+  _ fd: Int32, _ buf: UnsafeMutableRawPointer!, _ nbytes: CUnsignedInt
+) -> CInt {
+  _read(fd, buf, nbytes)
+}
+#else
 public func system_read(
   _ fd: Int32, _ buf: UnsafeMutableRawPointer!, _ nbyte: Int
 ) -> Int {
   read(fd, buf, nbyte)
 }
+#endif
 
 public func system_pread(
   _ fd: Int32, _ buf: UnsafeMutableRawPointer!, _ nbyte: Int, _ offset: off_t
 ) -> Int {
+#if os(Windows)
+  let handle: intptr_t = _get_osfhandle(fd)
+  if handle == /* INVALID_HANDLE_VALUE */ -1 { return Int(EBADF) }
+
+  // NOTE: this is a non-owning handle, do *not* call CloseHandle on it
+  let hFile: HANDLE = HANDLE(bitPattern: handle)!
+
+  var ovlOverlapped: OVERLAPPED = OVERLAPPED()
+  ovlOverlapped.OffsetHigh = DWORD(UInt32(offset >> 32) & 0xffffffff)
+  ovlOverlapped.Offset = DWORD(UInt32(offset >> 0) & 0xffffffff)
+
+  var nNumberOfBytesRead: DWORD = 0
+  if !ReadFile(hFile, buf, DWORD(nbyte), &nNumberOfBytesRead, &ovlOverlapped) {
+    let _ = GetLastError()
+    // TODO(compnerd) map windows error to errno
+    return Int(-1)
+  }
+  return Int(nNumberOfBytesRead)
```

Comment:
Could you explain more what is going on here? Does Windows not support pread/pwrite? We will have to decide on a case-by-case basis what should be handled by availability and what should be handled by implementing the functionality ourselves (but this doesn't look too bad). This is definitely worthy of being extracted into a helper function and given a good name (and comments!) since it's non-obvious to a non-Windows contributor what's going on.

Reply:
Yes, there is no `pread` or `pwrite` function that is available in the C library.  However, the operation itself _is_ supported, but you need to use the Win32 APIs to do that.  Basically, I have to convert the file descriptor to the kernel/win32 representation (`HANDLE`) and then I can perform the operation via the Win32 API (`ReadFile` or `WriteFile`) as I don't see the value in going directly to the syscall interface (`NtReadFile` or `NtWriteFile`) as they are more complicated to use (it would directly map to the kernel call `ZwReadFile` or `ZwWriteFile`).